### DISCUSSION TITLE
Meanings of "base", "side #1", "side #2" in conflict markers?

### DISCUSSION THREAD

**Discussion by wolfv6** [jj-vcs/jj#4151] [Q&A]
What do `base`, `side #1`, and `side #2` mean in the context of conflict markers?
Here is the content of an example file after jj rebase:
```
<<<<<<< Conflict 1 of 1
%%%%%%% Changes from base to side #1
-b1
+++++++ Contents of side #2
b1
b2
>>>>>>> Conflict 1 of 1 ends
```
I am working through the jj tutorial:
https://martinvonz.github.io/jj/latest/tutorial/#conflicts
https://martinvonz.github.io/jj/latest/conflicts/ > Conflict markers

Here is the full rebase example:
```
$ jj log
@  nzsnuzll wolf 2024-07-24 05:13:58 36cdb4ef
│  C
◉  yompupqv wolf 2024-07-24 05:13:52 7b4562ba
│  B2
◉  luuolwmk wolf 2024-07-24 05:13:03 e4801ddc
│  B1
◉  kuruumvk wolf 2024-07-24 05:12:41 127666c6
│  A
│ ◉  nvzmzxyo wolf 2024-07-24 04:38:11 ce55764b
├─╯  Say goodbye
│ ◉  tpstlust support+octocat@github.com 2018-05-10 11:55:19 octocat-patch-1@origin b1b3f972
├─╯  sentence case
│ ◉  kowxouwz octocat@nowhere.com 2014-06-10 16:22:26 test@origin b3cbd5bb
├─╯  Create CONTRIBUTING.md
◉  orrkosyo octocat@nowhere.com 2012-03-06 17:06:50 master 7fd1a60b
│  (empty) Merge pull request #6 from Spaceghost/patch-1
~
$ jj rebase -s y -d ku
Rebased 2 commits
New conflicts appeared in these commits:
  nzsnuzll 5b5d60cf (conflict) C
  yompupqv 82dc005e (conflict) B2
To resolve the conflicts, start by updating to the first one:
  jj new yompupqvrpqx
Then use `jj resolve`, or edit the conflict markers in the file directly.
Once the conflicts are resolved, you may want inspect the result with `jj diff`.
Then run `jj squash` to move the resolution into the conflicted commit.
Working copy now at: nzsnuzll 5b5d60cf (conflict) C
Parent commit      : yompupqv 82dc005e (conflict) B2
Added 0 files, modified 1 files, removed 0 files
There are unresolved conflicts at these paths:
file1    2-sided conflict
$ jj log
@  nzsnuzll wolf 2024-07-24 05:16:32 5b5d60cf conflict
│  C
◉  yompupqv wolf 2024-07-24 05:16:32 82dc005e conflict
│  B2
│ ◉  luuolwmk wolf 2024-07-24 05:13:03 e4801ddc
├─╯  B1
◉  kuruumvk wolf 2024-07-24 05:12:41 127666c6
│  A
│ ◉  nvzmzxyo wolf 2024-07-24 04:38:11 ce55764b
├─╯  Say goodbye
│ ◉  tpstlust support+octocat@github.com 2018-05-10 11:55:19 octocat-patch-1@origin b1b3f972
├─╯  sentence case
│ ◉  kowxouwz octocat@nowhere.com 2014-06-10 16:22:26 test@origin b3cbd5bb
├─╯  Create CONTRIBUTING.md
◉  orrkosyo octocat@nowhere.com 2012-03-06 17:06:50 master 7fd1a60b
│  (empty) Merge pull request #6 from Spaceghost/patch-1
~
$ cat file1
a
<<<<<<< Conflict 1 of 1
%%%%%%% Changes from base to side #1
-b1
+++++++ Contents of side #2
b1
b2
>>>>>>> Conflict 1 of 1 ends
```

**Comment by ilyagr** [comment]
Thanks! We haven't updated this docs for a while. Instead of answering right here, I might ask you to look at a PR updating the docs.

In short, assuming you follow what sides of a conflict are, you probably would like to know which side of the conflict corresponds to the commit being rebased. Currently, you have to guess, all that is guaranteed is side \#1 corresponds consistently to one of the two sides per-file. See also https://github.com/martinvonz/jj/issues/1176.

If that sounds cryptic, feel free to ask follow-up questions or wait until I hopefully make the PR updating the docs.

**Reply by wolfv6** [reply]
I looked at #1176 and still don't understand.  I probably need an example with explanation.  I look forward to seeing the updated docs.

**Reply by ilyagr** [reply]
PTAL at https://github.com/martinvonz/jj/pull/4168 and especially the second commit, also viewable at https://github.com/ilyagr/jj/commit/conflictdoc. Follow-up questions are welcome; I don't know whether it addresses your questions.

**Reply by amiryal** [reply]
@wolfv6: Do you find [this explanation](https://github.com/martinvonz/jj/pull/4168#discussion_r1693909329) helpful?

> We can resolve the conflict by manually making the changes (as indicated by sections marked by `%%%%%%%`) in the section marked by `+++++++`, and then removing all extraneous sections and markers.

So, in your case from the OP, `%%%%%%% Changes from base to side #1` indicate to remove `b1` that was there by itself and `+++++++ Contents of side #2` show `b1` with another line (`b2`) after it. It really depends on the wider context, so there is no definite “correct” answer, but the naïve resolution would be to change `+++++++ Contents of side #2` by removing `b1` and leaving `b2`, and thus the final contents after removing all extraneous sections and markers would be:
```
a
b2
```

**Reply by wolfv6** [reply]
Thanks amiryal.  It's starting to make more sense now.

**Comment by wolfv6** [comment]
Thank you for your patience.  I think I understand now.
From the tutorial https://martinvonz.github.io/jj/latest/tutorial/#conflicts :

Log before rebase:
```
C (b2)
|
B2 (b2)
|
B1 (b1)
|
A (a)
|
|/
|/
master
|
~
```
`jj rebase -s B2 -d A`

Log after rebase:
```
C (b2)
|
B2 (b2)
|
| B1 (b1)
|/
A (a)
|
|/
|/
master
|
~
```
file1 after rebase:
```
$ cat file1
<<<<<<< Conflict 1 of 1
%%%%%%% Changes from base to side #1
-b1
+a
+++++++ Contents of side #2
b2
>>>>>>> Conflict 1 of 1 ends
```
Are these statements about the above tutorial example correct?:

Base is the master branch.

Side #1 is destination branch and contains change A.
"-b1" means 'b1' was removed from history and is no longer a conflict.
"+a"  means 'a' is in conflict with side #2.

Side #2 is source branch and contains changes B1, B2, C.
"b2" is the snap shot of B2.

**Comment by wolfv6** [comment]
In #1176, the notation "conflict (B + (C - A))" seems like it might be a useful way to think of conflicts.
What do the '+' and '-' mean?
How does the notation work for this example from #1176?:
```
# Initial log
B
|
| C
|/
A

# Rebase C onto B:
C conflict (B + (C - A))
|
B
|
A
```
